packages/api/src/graphql.js

```diff
@@ -1,111 +1,113 @@
 "use strict";
 
+const { defaultFieldResolver } = require("graphql");
 const { ApolloServer, makeExecutableSchema } = require("apollo-server-express");
 const { SchemaDirectiveVisitor } = require("graphql-tools");
 
 const data = [
   // Returned regardless of auth roles
   {
     productId: 'Product 0',
-    cost: 0,
+    cost: 1.23,
     isSensitive: false,
     date: (new Date()).toString(),    
   },
   // Returned iff user has role 'Product 1'
   {
     productId: 'Product 1',
-    cost: 1,
+    cost: 3.45,
     isSensitive: true,
     date: (new Date()).toString(),    
   },
   // Returned iff user has role 'Product 2'
   {
     productId: 'Product 2',
-    cost: 2,
+    cost: 2.34,
     isSensitive: true,
     date: (new Date()).toString(),    
   }      
 ];
 
-// const RestrictedProductInformationDirective 
-
-class RestrictedProductInformationDirective extends SchemaDirectiveVisitor {
-  visitObject(type) {
-    const { productIdField, sensitiveFlagField, sensitiveDataField } = this.args;
-    const fields = type.getFields();
-    const hasAuthLogic = !!(productIdField && sensitiveFlagField && sensitiveDataField);
-    const hasAuthFields = !!(
-      Object.keys(fields).includes(productIdField) && 
-      Object.keys(fields).includes(sensitiveFlagField) &&
-      Object.keys(fields).includes(sensitiveDataField)
-    )
-
-    if (!hasAuthLogic) {
-      throw Error('RestrictedProductInformationDirective: missing required arguments');
-    }
-
-    if (!hasAuthFields) {
-      throw Error('RestrictedProductInformationDirective: arguments not in object fields');
-    }
-
-    const originalSensitiveDataFieldResolver = fields[sensitiveDataField].resolve;
-
-    // eslint-disable-next-line max-params
-    fields[sensitiveDataField].resolve = async function(parent, args, context, ...rest) {
-      const productField = parent[productIdField];
-      const isSensitive = parent[sensitiveFlagField];
-      const productRoles = context.me.productRoles;
-      const originalResult = originalSensitiveDataFieldResolver(parent, args, context, ...rest);
-      if (isSensitive) {
-        return productRoles.includes(productField) ? originalResult : null;
-      }
-      return originalResult;
-    }
+class Product {
+	constructor(productData) {
+		Object.assign(this, productData);
+	}
+
+	isRestricted() {
```

Comment:
With domain objects and duck typing, we get flexibility with regard to which properties in a given object flag something as restricted without sacrificing the generic-ness of the directive code.

Reply:
This also ensures that if we have an error where a particular object does not quack to `isRestricted` it will be an error at development time rather than a potential access violation in a deployed environment.